### PR TITLE
[unit-tests] Added unit tests for collaborators helper

### PR DESCRIPTION
**Pull request by kiranparajuli589**
## Description
- added unit tests for files package helper `collaborators.js`

## Related Issue
- Part of #5236 

## How Has This Been Tested?
- 🤖 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Review by saw-jan — COMMENTED** [review]
LGTM
you can use `unit-tests-only` in title. The CI build time will be faster

**Review by dpakach — COMMENTED** [review]
LGTM!

**Comment by kiranparajuli589** [comment]
> LGTM
> you can use `unit-tests-only` in title. The CI build time will be faster

isn't the script doing that job automatically?
see acceptance test suites just running for less than a minute https://drone.owncloud.com/owncloud/web/18136

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      update: 2,
+      create: 4,
+      delete: 8,
+      share: 16,
+      all: 31
+    })
+  })
+
+  describe('roleToBitmask function', () => {
+    describe('if additionalPermissions', () => {
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:21]: is this needed? I'm confused about this sub-describe

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      it('should return the sum of role permissions and additional permissions bitmasks if the given role has provided additional permission', () => {
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:22]: create more variations of the test, e.g. also test corner cases, invalid input, interesting combinations of the permissions etc.

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                share: true
+              },
+              permissions: ['read']
+            },
+            ['share', 'update']
+          )
+        ).toBe(17)
+      })
+    })
+    it('should return the sum of role permission bitmask if no additionalPermissions are given', () => {
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:36]: create more variations of the test, e.g. also test corner cases, invalid input, interesting combinations of the permissions etc.

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            permissions: ['read', 'create']
+          },
+          []
+        )
+      ).toBe(5)
+    })
+  })
+
+  describe('bitmaskToRole function', () => {
+    it('should return advancedRole if rolePermissionsBitmask or roleBasicPermissionsBitmask does not equal provided bitmask', () => {
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:51]: this description and the comment bellow does refer to internal variable names of the function, we should not do that, those things might be refactored. We only want to test and rely on the external API of the function, this applies also to the description, otherwise if those variables will be refactored in the function the test might still pass, but neither the description nor the comment would make any sense.

Also more variations of the test are needed. What about if its not a folder, different bitmask, invalid values, etc.

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      // two roles with following permissions are available from roleDefinition helper
+      // - viewer with permissions ['read']
+      // - editor with permissions ['read', 'update']
+      // with any of the role, the sum of permission bitmask does not matches provided bitmask
+      // so the function will return advancedRole
+      expect(bitmaskToRole(17, false, false)).toMatchObject({
+        name: 'advancedRole'
+      })
+    })
+    it('should return currentRole if rolePermissionsBitmask or roleBasicPermissionsBitmask equals provided bitmask', () => {
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:62]: same comment as above

**Review by individual-it — CHANGES_REQUESTED** [review]

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    it.each(['17', 17.2, '17.8', ['17.8']])('should parse provided bitmask to integer', bitmask => {
+      expect(bitmaskToRole(bitmask, false, true)).toMatchObject({
+        name: 'viewer'
+      })
+    })
+    it.each([true, 0, {}])('should return advanced role for invalid bitmask values', bitmask => {
+      expect(bitmaskToRole(bitmask, false, true)).toMatchObject({
+        name: 'advancedRole'
+      })
+    })
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:229]: @kulmann is that what we want? or should there be an error thrown for invalid bitmasks?

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    it('should return false if given bitmask does not equal permissionBitmask', () => {
+      expect(checkPermission(permissionsBitmask.read, 'update')).toBeFalsy()
+    })
+    it('should return false if given permission is not found in permissions bitmask', () => {
+      expect(checkPermission(1, 'modify')).toBeFalsy()
+    })
+    it.each([
+      // seems an edge case to me, should this be fixed?
+      // if permission is valid and bitmask is `true` function returns truthy value
+      { bitmask: true, permission: 'read', expected: true },
```
Comment (individual-it) [inline packages/web-app-files/tests/unit/helpers/collaborators.spec.js:250]: @kulmann what do you think?

**Comment by kiranparajuli589** [comment]
@kulmann please have a look. we need your opinion here.

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js (hunk 7 above)
Comment (kulmann) [reply packages/web-app-files/tests/unit/helpers/collaborators.spec.js:250]: Hmm. we should do two things:
1. convert the helper to typescript and allow string and number as types for the bitmask
2. as we won't have strict type checks in the near future it's probably still a good idea to add a typecheck into the code that only string and number are allowed. i.e. return `false` when the bitmask is neither a string nor a number.

If you want to you could add the type check (the second bullet) to the helper and adapt the test here accordingly. Otherwise please open a ticket and we can do it then. Thanks for the pointer! 👍

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js (hunk 6 above)
Comment (kulmann) [reply packages/web-app-files/tests/unit/helpers/collaborators.spec.js:229]: advanced role for the `0` bitmask would be fine. That bitmask doesn't make too much sense, but it's still valid. For the wrong types there should be some error.

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js (hunk 7 above)
Comment (kiranparajuli589) [reply packages/web-app-files/tests/unit/helpers/collaborators.spec.js]: i've added a type check for the functions, please reveiw again

File: packages/web-app-files/tests/unit/helpers/collaborators.spec.js (hunk 6 above)
Comment (kiranparajuli589) [reply packages/web-app-files/tests/unit/helpers/collaborators.spec.js]: thrown error if wrong types are supplied. please review

File: packages/web-app-files/src/helpers/collaborators.js [hunk 8]
```diff
@@ -44,6 +44,10 @@ export function roleToBitmask(role, additionalPermissions = []) {
  * @returns {object} Role mapped to the bitmask
  */
 export function bitmaskToRole(bitmask, isFolder, allowSharePerm) {
+  if (typeof bitmask !== 'number')
```
Comment (kulmann) [inline packages/web-app-files/src/helpers/collaborators.js:47]: string is also allowed here

File: packages/web-app-files/src/helpers/collaborators.js [hunk 9]
```diff
@@ -44,6 +44,10 @@ export function roleToBitmask(role, additionalPermissions = []) {
...
+    throw new Error(
+      'TypeError: bitmask is expected to be a number but was found: ' + typeof bitmask
```
Comment (kulmann) [inline packages/web-app-files/src/helpers/collaborators.js:49]:
```suggestion
      'TypeError: bitmask is expected to be a number or stringified number but was found: ' + typeof bitmask
```

**Review by kulmann — CHANGES_REQUESTED** [review]

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingPublicManagement</strong> https://drone.owncloud.com/owncloud/web/18727/34/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

File: packages/web-app-files/src/helpers/collaborators.js (hunk 9 above)
Comment (kiranparajuli589) [reply packages/web-app-files/src/helpers/collaborators.js]: fixed

File: packages/web-app-files/src/helpers/collaborators.js (hunk 8 above)
Comment (kiranparajuli589) [reply packages/web-app-files/src/helpers/collaborators.js]: fixed

**Comment by ownclouders** [comment]
Results for <strong>oC10iPhone2</strong> https://drone.owncloud.com/owncloud/web/18741/46/1
The following scenarios passed on retry:
- webUITrashbinRestore/trashbinRestore.feature:14

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingExternalRoot</strong> https://drone.owncloud.com/owncloud/web/18741/40/1
The following scenarios passed on retry:
- webUISharingExternalToRoot/federationSharing.feature:356

**Comment by ownclouders** [comment]
Results for <strong>oCISFiles1</strong> https://drone.owncloud.com/owncloud/web/19070/52/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingInternalUsersSharingIndicator</strong> https://drone.owncloud.com/owncloud/web/19125/28/1
The following scenarios passed on retry:
- webUISharingInternalUsersToRootSharingIndicator/shareWithUsers.feature:98

**Comment by ownclouders** [comment]
Results for <strong>oC10iPhone1</strong> https://drone.owncloud.com/owncloud/web/19125/45/1
The following scenarios passed on retry:
- webUIDeleteFilesFolders/deleteFilesFolders.feature:59

**Comment by pascalwengerter** [comment]
@kulmann @individual-it could you re-review? Seems like your reviews have been addressed

**Comment by individual-it** [comment]
@kiranparajuli589 lint is not passing

**Review by individual-it — APPROVED** [review]
except of lint it looks good to me

**Comment by individual-it** [comment]
lint is passing now @kulmann please review again

**Comment by ownclouders** [comment]
Results for <strong>oC10CreateDelete</strong> https://drone.owncloud.com/owncloud/web/20340/12/1
The following scenarios passed on retry:
- webUIDeleteFilesFolders/deleteFilesFolders.feature:59

**Comment by individual-it** [comment]
blocked waiting for review @kulmann @pascalwengerter

**Comment by ownclouders** [comment]
Results for <strong>oCISBasic</strong> https://drone.owncloud.com/owncloud/web/20720/51/1
The following scenarios passed on retry:
- webUIAccount/accountInformation.feature:34
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by kiranparajuli589** [comment]
ci failed at `smoke-tests OCIS` https://drone.owncloud.com/owncloud/web/20722/11/1
ci restarted ♻️

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingIntUsers1</strong> https://drone.owncloud.com/owncloud/web/20730/59/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by pascalwengerter** [comment]
Has this branched been rebased lately? Acceptance test failures for those (small) changes seem suspicious

**Comment by pascalwengerter** [comment]
@phil-davis the `[unit-tests]` shouldn't make any difference since the CI now figures out what to run based on changed files. Kiran added some logic to a JS helper from the actual code so now the acceptance tests are being run (and can fail), too, in addition to the unit test suite

**Comment by individual-it** [comment]
@pascalwengerter the branch is up-to-date
something is strange in ocis
can you make any sense out of the logs?
```
{"level":"error","service":"storage","pkg":"rgrpc","traceid":"00000000000000000000000000000000","error":"symlink ../b83bc447-9e07-48a9-9ecd-cd6f9c137aeb /srv/app/tmp/ocis/storage/users/nodes/root/ddc2004c-0977-11eb-9d3f-a793888cd0f8: file exists","time":"2021-12-08T08:54:59Z","message":"storageprovider: error calling CreateHome of storage driver"}
{"level":"error","service":"proxy","error":"gateway: grpc failed with code CODE_INTERNAL","time":"2021-12-08T08:54:59Z","message":"error when calling Createhome"}
{"level":"error","service":"ocs","error":"{\"code\":404,\"detail\":\"The requested user could not be found\",\"status\":\"Not Found\"}","userid":"Brian","time":"2021-12-08T08:54:59Z","message":"could not delete user"}

**Comment by individual-it** [comment]
^ Those logs are actually fine, the tests try to delete the user before creating them, so in the first iteration the user does not exist

**Comment by individual-it** [comment]
This looks like the real issue:
```
{"level":"error","service":"web","error":"listen tcp 127.0.0.1:33579: bind: address already in use","transport":"http","time":"2021-12-08T08:54:33Z","message":"Failed to start server"}
```

**Comment by kiranparajuli589** [comment]
@individual-it @pascalwengerter Just for debugging _(the last commit will be reverted)_, I've used the previous commit ID for ocis before https://github.com/owncloud/web/pull/6093/files.
let's see how the tests behave with the previous version of ocis.

**Comment by kiranparajuli589** [comment]
@individual-it @pascalwengerter Just for debugging _(the last commit will be reverted)_, I've used the previous commit ID for ocis before https://github.com/owncloud/web/pull/6093/files.
let's see how the tests behave with the previous version of ocis.

**Comment by kiranparajuli589** [comment]
@individual-it @pascalwengerter Just for debugging _(the last commit will be reverted)_, I've used the previous commit ID for ocis before https://github.com/owncloud/web/pull/6093/files.
let's see how the tests behave with the previous version of ocis.

**Comment by pascalwengerter** [comment]
The tested helper will get deleted in https://github.com/owncloud/web/pull/6039/files, closing this here and stopping the CI efforts. Sorry for the late realization

**Comment by pascalwengerter** [comment]
@kiranparajuli589 type-checking the bitmask was the right idea btw, only that we've refactored some code and moved it to TypeScript where this happens automatically 😉
